File: python/lda.py
import numpy as np
import pandas as pd

# Import dataset
df = pdf.read_csv('data.csv')
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

# Split into training and test sets
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Apply LDA
# By default, it reduces to k-1 components, 
# where k is the number of classes
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda = LinearDiscriminantAnalysis(n_components=None)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

# LDA can classify data using posteriors
y_pred = np.max(lda.decision_function(), 1)

File: r/lda.r
# Import dataset
df = read.csv('data.csv')

# Create Boolean vector with random split
library(caTools)
split = sample.split(df$y, SplitRatio=0.8)

# Split into training and test set
train = subset(df, split == TRUE)
test = subset(df, split == FALSE)

# Feature scaling
# Pick only numerical columns
train[, -ncol(train)] = scale(train[, -ncol(train)])
test[, -ncol(test)] = scale(test[, -ncol(test)])

# Functions to support Venables and Ripley
# "Modern Applied Staitstics with S" (2002)
install.packages('MASS')
library(MASS)

# Apply LDA
lda = lda(groups ~ ., data=train)
train_lda = predict(lda, train)
train = as.data.frame(cbind(train_lda$x, train_lda$class))
test_lda = predict(lda, test)
test = as.data.frame(cbind(test_lda$x, test_lda$class))

# LDA can classify data using posteriors
y_pred = apply(test_lda$posterior, 1, which.max)
